File: bindings/javascript/src/safe_password.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafePassword - Password validation that cannot crash.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 * Note: proven_password_validate returns a PasswordResult struct which
 * requires buffer-based marshaling. proven_password_is_common is directly
 * callable.
 * @module
 */

import { getLib } from './ffi.js';
import { ok, err } from './result.js';

/**
 * Password strength levels.
 * @readonly
 * @enum {number}
 */
export const PasswordStrength = Object.freeze({
  VERY_WEAK: 0,
  WEAK: 1,
  FAIR: 2,
  STRONG: 3,
  VERY_STRONG: 4,
});

/**
 * Safe password operations backed by formally verified Idris 2 code.
 */
export class SafePassword {
  /**
   * Check if a password is in the common passwords list.
   *
   * @param {string} password - The password to check.
   * @returns {boolean} True if the password is common.
   */
  static isCommon(password) {
    const symbols = getLib();
    const bytes = new TextEncoder().encode(password);
    return symbols.proven_password_is_common(bytes, bytes.length);
  }
}

File: bindings/zig/src/safe_password.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafePassword - FFI bindings to libproven password operations.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Password strength levels.
pub const PasswordStrength = enum(c_int) {
    very_weak = c.PROVEN_PASSWORD_VERY_WEAK,
    weak = c.PROVEN_PASSWORD_WEAK,
    fair = c.PROVEN_PASSWORD_FAIR,
    strong = c.PROVEN_PASSWORD_STRONG,
    very_strong = c.PROVEN_PASSWORD_VERY_STRONG,
};

/// Password validation result.
pub const PasswordResult = struct {
    strength: PasswordStrength,
    has_lowercase: bool,
    has_uppercase: bool,
    has_digit: bool,
    has_special: bool,
    length: usize,
};

/// Validate password strength via libproven.
pub fn validate(password: []const u8) PasswordResult {
    const result = c.proven_password_validate(password.ptr, password.len);
    return PasswordResult{
        .strength = @enumFromInt(result.strength),
        .has_lowercase = result.has_lowercase,
        .has_uppercase = result.has_uppercase,
        .has_digit = result.has_digit,
        .has_special = result.has_special,
        .length = result.length,
    };
}

/// Check if password is in common passwords list via libproven.
pub fn isCommon(password: []const u8) bool {
    return c.proven_password_is_common(password.ptr, password.len);
}

test "validate" {
    const result = validate("MyP@ssw0rd!");
    try std.testing.expect(result.has_uppercase);
    try std.testing.expect(result.has_lowercase);
    try std.testing.expect(result.has_digit);
    try std.testing.expect(result.has_special);
}

test "isCommon" {
    try std.testing.expect(isCommon("password"));
    try std.testing.expect(!isCommon("xK9$mZ2!vQ7@nW4#"));
}
